climada/hazard/base.py

```diff
@@ -1393,3 +1442,167 @@ def _read_att_excel(self, file_name, var_names):
+    units = {haz.units for haz in haz_list}
+    if len(units) > 1:
+        raise TypeError("The haz_list contains hazards with different"
+                        "units %f. The hazards are incompatible and"
+                        "cannot be concatenated.", units)
+
+    #Define comon centroids
+    if centroids is None:
+        centroids = join_centroids(haz_list)
+
+    haz_concat = Hazard()
+    haz_concat.units = haz_list[0].units
+    haz_concat.centroids = centroids
+
+    #Indices for mapping matrices onto common centroids
+    hazcent_in_cent_idx_list = [
+        u_coord.assign_coordinates(haz.centroids.coord, centroids.coord)
+        for haz in haz_list
+        ]
+
+    #Concatenate attributes - hazards are assumed to have the same attributes
+    for attr_name in vars(haz_list[0]).keys():
+        attr_val_list = [getattr(haz, attr_name) for haz in haz_list]
+        if isinstance(attr_val_list[0], sparse.csr.csr_matrix):
+            matrix = (
+                _map_matrix_to_centroids(matrix,
+                                        centroids,
+                                        cent_idx
+                                        )
+                for matrix, cent_idx in zip(attr_val_list, hazcent_in_cent_idx_list)
+                )
+            setattr(haz_concat, attr_name, sparse.vstack(matrix, format='csr'))
+        elif (isinstance(attr_val_list[0], np.ndarray)
+              and attr_val_list[0].ndim == 1):
+            setattr(haz_concat, attr_name, np.hstack(attr_val_list))
+        elif isinstance(attr_val_list[0], list):
+            setattr(haz_concat, attr_name, sum(attr_val_list, []))
+        elif isinstance(attr_val_list[0], TagHazard):
+            for tag in attr_val_list:
+                if tag is not haz_concat.tag: haz_concat.tag.append(tag)
+
+    haz_concat.sanitize_event_ids()
+
+    return haz_concat
+
+def _map_matrix_to_centroids(sp_matrix, centroids, cent_idx):
+    """
+    Map a sparse matrix (e.g. hazard intensity or fraction) onto centroids.
+
+    Parameters
+    ----------
+    matrix: scipy.sparse.crs_matrix()
+        Input hazard. haz.centroids must be a subset of centroids
+    centroids: climada.hazard.Centroids()
+        Centroids on which to map the haz.intensity matrix
+    cent_idx: numpy.ndarray
+        Array of indices for sp_matrix centroids.
+
+    Returns
+    -------
+    mapped_matrix: scipy.sparse.csr_matrix()
+        Sparse fraction matrix of dimension haz.size X centroids.size
+    """
+    n_events = sp_matrix.shape[0]
+    chunks = min(int(n_events / 100), 1000)
+    counter = np.array_split(np.arange(n_events), chunks + 1)
+    mapped_matrix = sparse.csr_matrix([])
+    events = []
+
+    for row in counter:
+        for n in row:
+            ar = np.zeros(int(centroids.size))
+            ar[cent_idx] = sp_matrix.getrow(n).todense()
+            events.append(sparse.csr_matrix(ar))
+        mapped_matrix = sparse.vstack(events)
```

Comment:
My sparse matrix know-how isn't the best, so I wanted to ask if there's a neater way to do this, since you're essentially doing a column permutation. Could you use a permutation matrix instead (addding any necessary all-zero columns as well)? Is there a sparse-specific way to do this?

Reply:
In fact, this is a "one-liner" with the CSR format because the number and values of non-zero entries don't change:
```
sparse.csr_matrix((sp_matrix.data, cent_idx[sp_matrix.indices], sp_matrix.indptr),
                  shape=(sp_matrix.shape[0], centroids.size))
```
Since this functionality is used only once, I wouldn't bother defining a new function for it.

Reply:
Having it as a separate function makes it quite a bit more readable, and it might be use elsewhere. But, I am also fine with having this one-liner too.